config/jobs/kubernetes/sig-network/sig-network-misc.yaml

```diff
@@ -242,3 +242,34 @@ periodics:
       - --test_args=--ginkgo.focus=\[Feature:Ingress\] --ginkgo.skip=\[Unreleased\] --minStartupPods=8
       - --timeout=300m
       image: gcr.io/k8s-testimages/kubekins-e2e:v20181017-0323abc4c-master
+
+- interval: 6h
+  name: ci-kubernetes-e2e-kubeadm-dind-ipv6
+  labels:
+    preset-service-account: "true"
+  spec:
+    containers:
+    - args:
+      - --tmpfs=/docker-graph:exec
+      - --repo=k8s.io/$(REPO_NAME)=$(PULL_REFS)
```

Comment:
I think you don't need k8s repo here, as we are using latest built binary now

Reply:
K-D-C (and the kubeadm-dind kubetest deployer) currently does its own build of Kubernetes binaries from a Kubernetes tree pulled using REPO_NAME and PULL_REFS. I was planning on (eventually) switching this to use the latest built binaries, either pulled down from Google Storage, or if kubetest is already pulling the latest build binaries into the host Prow container, volume mounting those binaries in each of the node containers. The latter method would require changes to both K-D-C scripts and to the kubeadm-dind kubetest deployer. Can we do this as a followup change?

Reply:
@krzyzacy - Is the K-D-C build of binaries (vs. using latest built binaries) a showstopper, or can we make this change as a followup? K-D-C does have the capability to pull down binaries from Google Storage, but to use that I'd have to modify the kubeadm-dind deployer in kubetest to get passed some build references (or a URL) to know where to pull binaries from GS. Or I could change KDC so that the container nodes are able to read binaries on the host prow container (e.g. bind mount). (The last option would take some time to get K-D-C change merged on master and get bumped to the K-D-C stable branch.)

Reply:
sorry, previous github notifications are messed up so I didn't see your previous comment

the `--extract=ci/latest` downloads the k8s tarball built from latest CI, I don't see here how you are using the k8s source repo, do you want to call some extra build commands?

Reply:
@krzyzacy - The way it works now is that K-D-C must be run from a kubernetes source tree (/workspace/src/k8s.io/kubernetes/), and it does its own build of hyperkube and kubeadm from that source. So the --repo is intended to pull/populate that kubernetes source tree. The build that K-D-C is doing is superfluous (for periodic job at least, and probably the pre-submit?), but it will take some changes to K-D-C and the kubeadm-dind deployer to make use of the latest tarball.

Reply:
So your code can tell the difference between a k8s source and a k8s binary?

Your workspace will look like
```
/go
  /src
    /k8s.io
      /kubernetes # <---- k8s source from --repo
        /kubernetes #  <---- k8s binary from --extract=ci/latest, cwd
    /github.com
      /kubernetes-sigs
        /kubeadm-dind-cluster
```

Reply:
also this ($PULL_REFS) won't be resolved in periodics... If you really want a copy of source, you can use https://github.com/kubernetes/test-infra/blob/master/kubetest/main.go#L141?

Reply:
Ah, good to know. Is the "extract-source" usable for pre-submits as well? And if I understand correctly, with "extract-source" the source will be extracted to /go/src/k8s.io/kubernetes/ ?

Reply:
I don't think so... it will be next to the tarball... if you want to keep it in GOPATH then maybe just checkout master? or pin PULL_REF to a specific SHA?

Reply:
The source doesn't have to be in GOPATH, it could be anywhere. I just need to know where to 'cd ...' to in the prow container. So "next to the tarball" would be in /go/src/k8s.io/kubernetes/kubernetes?

Reply:
so in your case this will resolve to:

```
/go
  /src
    /github.com
      /kubernetes-sigs
        /kubeadm-dind-cluster
          /kubernetes #  <---- k8s binary + src
```

Reply:
(ref https://github.com/kubernetes/test-infra/blob/master/kubetest/extract_k8s.go#L235-L300)

Reply:
That snippet helps! I see it's relative to cwd (sorry, I missed the "cwd" reference in your earlier comment). I think cwd is /workspace at the time this runs. Let me run locally and confirm, thanks!!!

Reply:
--repo flag will change your cwd..

Reply:
@krzyzazy - I have a bit of a dilemma because of PR #9915 . This PR reverted [one of the changes that I had added](https://github.com/kubernetes/test-infra/pull/9915/files#diff-1348bfba2403286944b343e21e901c14L589). Looks like PR #9915 was needed because the kubernetes-anywhere had been using provider=local (though that's been temporarily? reverted with PR #9919), and they need to have GCE SSH populated.

The kubeadm-dind deployer also uses provider=local, and GCE SSH connectivity is not required. I'm not setting the GCE SSH key environment variables, so the kubernetes_e2e scenario is crashing in mode.add_gce_ssh().

I don't know if PR #9915 will eventually be reverted (e.g. if PR #9919 is permanent).

Should I just go ahead set GCE SSH environment (does the preset-k8s-ssh flag cover that?), even though I don't need GCE SSH connectivity? I can't simply leave args.gce_ssh to be undefined, since it gets defaulted to a non-nil value (e.g. the 'elif args.gce_ssh' conditional should always resolve to true, as far as I can tell.)

Reply:
yeah that's totally fine to include the ssh preset, just add a TODO or comment to follow up

Reply:
@krzyzacy - Thanks, I'll add the K8s ssh preset.
I'm trying to use the extract-source flag, but the extraction isn't finding Kubernetes source in the downloaded tarball: 
```
root@e06b4988ab17:/workspace/github.com/kubernetes-sigs/kubeadm-dind-cluster/kubernetes/kubernetes# ls
README-src.txt
root@e06b4988ab17:/workspace/github.com/kubernetes-sigs/kubeadm-dind-cluster/kubernetes/kubernetes# cat README-src.txt 
For build efficiency, the src was not included in this release.
The full source code can be viewed at
https://github.com/kubernetes/kubernetes/tree/e463d10935c4e4ef683e26c63a5994976addef11
root@e06b4988ab17:/workspace/github.com/kubernetes-sigs/kubeadm-dind-cluster/kubernetes/kubernetes# 
```
Is this normal/expected?

Reply:
But actual release versions are fairly infrequent, right? (Compared to a 6-hour periodic?)

I'm trying to figure out what it will take to have K-D-C bind-mount the Kubernetes binaries that are extracted in the Prow container... but that will take changes to K-D-C and the kubeadm-dind deployer.

For the near term, is there an easy way to read the README-src.txt file somewhere in Google storage, and then git clone the Kubernetes tree using the github URL in that README-src.txt file, or something of that sort? I thought that there used to be a place in GS to read that URL directly without extracting a tarball.

Reply:
Would it be worth considering adding fallback code for --extract-source such that if the source code isn't included in the release (i.e. the ./kubernetes directory doesn't have a valid tree), it reads the README-src.txt file, gets the github location for the tree, and clones the source?

Having the extraction happening in /workspace/github.com/kubernetes-sigs/kubeadm-dind-cluster is awkward... I assume that's happening because of the --repo flag? Do you know if there's an easy way to have kubetest extract in another directory, e.g. /workspace?

Also, it doesn't appear that --extract is causing the server binaries (e.g. hyperkube and kubeadm, the 2 binaries I would need) to be extracted from the tarball. I suppose the deployer would have to extract these server binaries?

Reply:
Yes we are planning to move all extract logic to $HOME/.kubetest instead of cwd...

However we are going to write a v2 version instead of ducttape the existing one  :-)

> Also, it doesn't appear that --extract is causing the server binaries (e.g. hyperkube and kubeadm, the 2 binaries I would need) to be extracted from the tarball. I suppose the deployer would have to extract these server binaries?

hummm `--extract` basically calls into [`get-kube.sh`](https://github.com/kubernetes/kubernetes/blob/master/cluster/get-kube.sh)... you can check if that supports extra binaries...

Reply:
get-kube.sh calls get-kube-binaries.sh, and that script is downloading but not extracting server tar:
```
if "${DOWNLOAD_SERVER_TAR}"; then
  download_tarball "${KUBE_ROOT}/server" "${SERVER_TAR}"
fi

if "${DOWNLOAD_CLIENT_TAR}"; then
  download_tarball "${KUBE_ROOT}/client" "${CLIENT_TAR}"
  extract_arch_tarball "${KUBE_ROOT}/client/${CLIENT_TAR}" "${CLIENT_PLATFORM}" "${CLIENT_ARCH}"
fi

if "${DOWNLOAD_TESTS_TAR}"; then
  download_tarball "${KUBE_ROOT}/test" "${TESTS_TAR}"
  echo "Extracting ${TESTS_TAR} into ${KUBE_ROOT}"
  # Strip leading "kubernetes/"
  tar -xzf "${KUBE_ROOT}/test/${TESTS_TAR}" --strip-components 1 -C "${KUBE_ROOT}"
fi
```

Reply:
hummm... hummm... hummm... maybe you can extract that in your deployer test setup... welcome to the bash :fire:

Reply:
Yeah, that's where I was headed... but maybe do the server binary extraction in the NewDeployer() or in the Up() method (golang). Then change K-D-C to have the node containers bind mount that kube binary directory in the prow container. Might work. I'd want to use a temporary special branch of K-D-C for this, since waiting for this to trickle over to the K-D-C stable branch will take some time.

Reply:
@krzyzacy - Okay, the extraction of downloaded Kubernetes server binaries tar file was added for the kubeadmdind deployer with PR #10071. So this config change depends upon that test-infra PR, as well as [K-D-C PR # 247](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/247).

Until  [K-D-C PR # 247](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/247) is merged and the K-D-C stable branch is bumped up to incorporate this fix, this config uses a temporary fork of K-D-C that has the fix for [K-D-C PR # 247](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/247).